File: android-app/app/src/main/java/com/chainlesschain/android/remote/commands/UserBrowserCommands.kt
package com.chainlesschain.android.remote.commands

import com.chainlesschain.android.remote.client.RemoteCommandClient
import kotlinx.serialization.Serializable
import javax.inject.Inject
import javax.inject.Singleton

/**
 * 用户浏览器控制命令 API
 *
 * 通过 Chrome DevTools Protocol (CDP) 连接和控制用户已安装的真实浏览器
 * 支持 Chrome、Edge、Brave 等基于 Chromium 的浏览器
 */
@Singleton
class UserBrowserCommands @Inject constructor(
    private val client: RemoteCommandClient
) {
    // ==================== 连接管理 ====================

    /**
     * 查找可用浏览器
     *
     * @return 可用浏览器列表
     */
    suspend fun findBrowsers(): Result<FindBrowsersResponse> {
        return client.invoke("userBrowser.findBrowsers", emptyMap())
    }

    /**
     * 连接到用户浏览器
     *
     * @param browserType 浏览器类型（chrome/edge/brave）
     * @param port 调试端口
     * @param autoLaunch 是否自动启动浏览器
     */
    suspend fun connect(
        browserType: String = "chrome",
        port: Int? = null,
        autoLaunch: Boolean = true
    ): Result<ConnectBrowserResponse> {
        val params = mutableMapOf<String, Any>(
            "browserType" to browserType,
            "autoLaunch" to autoLaunch
        )
        port?.let { params["port"] = it }

        return client.invoke("userBrowser.connect", params)
    }

    /**
     * 断开浏览器连接
     */
    suspend fun disconnect(): Result<DisconnectBrowserResponse> {
        return client.invoke("userBrowser.disconnect", emptyMap())
    }

    /**
     * 获取连接状态
     */
    suspend fun getStatus(): Result<BrowserConnectionStatus> {
        return client.invoke("userBrowser.getStatus", emptyMap())
    }

    // ==================== 标签页管理 ====================

    /**
     * 列出所有标签页
     */
    suspend fun listTabs(): Result<UserBrowserTabsResponse> {
        return client.invoke("userBrowser.listTabs", emptyMap())
    }

    /**
     * 获取当前活动标签页
     */
    suspend fun getActiveTab(): Result<UserBrowserTabResponse> {
        return client.invoke("userBrowser.getActiveTab", emptyMap())
    }

    /**
     * 创建新标签页
     *
     * @param url 目标 URL（可选）
     * @param active 是否激活
     */
    suspend fun createTab(
        url: String? = null,
        active: Boolean = true
    ): Result<UserBrowserTabResponse> {
        val params = mutableMapOf<String, Any>("active" to active)
        url?.let { params["url"] = it }

        return client.invoke("userBrowser.createTab", params)
    }

    /**
     * 关闭标签页
     *
     * @param targetId 标签页 ID
     */
    suspend fun closeTab(targetId: String): Result<CloseTabResult> {
        return client.invoke("userBrowser.closeTab", mapOf("targetId" to targetId))
    }

    /**
     * 聚焦标签页
     *
     * @param targetId 标签页 ID
     */
    suspend fun focusTab(targetId: String): Result<FocusTabResult> {
        return client.invoke("userBrowser.focusTab", mapOf("targetId" to targetId))
    }

    // ==================== 导航 ====================

    /**
     * 导航到 URL
     *
     * @param targetId 标签页 ID
     * @param url 目标 URL
     */
    suspend fun navigate(targetId: String, url: String): Result<NavigateResult> {
        return client.invoke("userBrowser.navigate", mapOf(
            "targetId" to targetId,
            "url" to url
        ))
    }

    /**
     * 后退
     *
     * @param targetId 标签页 ID
     */
    suspend fun goBack(targetId: String): Result<NavigationResult> {
        return client.invoke("userBrowser.goBack", mapOf("targetId" to targetId))
    }

    /**
     * 前进
     *
     * @param targetId 标签页 ID
     */
    suspend fun goForward(targetId: String): Result<NavigationResult> {
        return client.invoke("userBrowser.goForward", mapOf("targetId" to targetId))
    }

    /**
     * 刷新页面
     *
     * @param targetId 标签页 ID
     * @param ignoreCache 是否忽略缓存
     */
    suspend fun refresh(
        targetId: String,
        ignoreCache: Boolean = false
    ): Result<NavigationResult> {
        return client.invoke("userBrowser.refresh", mapOf(
            "targetId" to targetId,
            "ignoreCache" to ignoreCache
        ))
    }

    // ==================== 页面操作 ====================

    /**
     * 执行 JavaScript（需要 Admin 权限）
     *
     * @param targetId 标签页 ID
     * @param script JavaScript 代码
     */
    suspend fun executeScript(
        targetId: String,
        script: String
    ): Result<ScriptExecutionResult> {
        return client.invoke("userBrowser.executeScript", mapOf(
            "targetId" to targetId,
            "script" to script
        ))
    }

    /**
     * 获取页面内容
     *
     * @param targetId 标签页 ID
     * @param format 格式（html/text）
     */
    suspend fun getPageContent(
        targetId: String,
        format: String = "html"
    ): Result<PageContentResult> {
        return client.invoke("userBrowser.getPageContent", mapOf(
            "targetId" to targetId,
            "format" to format
        ))
    }

    /**
     * 截图
     *
     * @param targetId 标签页 ID
     * @param format 格式（png/jpeg）
     * @param quality JPEG 质量（1-100）
     * @param fullPage 是否全页截图
     */
    suspend fun screenshot(
        targetId: String,
        format: String = "png",
        quality: Int? = null,
        fullPage: Boolean = false
    ): Result<UserBrowserScreenshotResult> {
        val params = mutableMapOf<String, Any>(
            "targetId" to targetId,
            "format" to format,
            "fullPage" to fullPage
        )
        quality?.let { params["quality"] = it }

        return client.invoke("userBrowser.screenshot", params)
    }

    // ==================== 用户数据 ====================

    /**
     * 获取书签
     *
     * @param folderId 文件夹 ID（可选）
     */
    suspend fun getBookmarks(folderId: String? = null): Result<UserBrowserBookmarksResponse> {
        val params = mutableMapOf<String, Any>()
        folderId?.let { params["folderId"] = it }

        return client.invoke("userBrowser.getBookmarks", params)
    }

    /**
     * 获取历史记录
     *
     * @param query 搜索关键词
     * @param maxResults 最大结果数
     * @param startTime 开始时间戳
     * @param endTime 结束时间戳
     */
    suspend fun getHistory(
        query: String? = null,
        maxResults: Int = 100,
        startTime: Long? = null,
        endTime: Long? = null
    ): Result<UserBrowserHistoryResponse> {
        val params = mutableMapOf<String, Any>("maxResults" to maxResults)
        query?.let { params["query"] = it }
        startTime?.let { params["startTime"] = it }
        endTime?.let { params["endTime"] = it }

        return client.invoke("userBrowser.getHistory", params)
    }
}

// ==================== 响应数据类 ====================

@Serializable
data class FindBrowsersResponse(
    val success: Boolean = true,
    val browsers: List<AvailableBrowser>
)

@Serializable
data class AvailableBrowser(
    val type: String,
    val name: String,
    val executablePath: String,
    val defaultPort: Int,
    val userDataDir: String? = null
)

@Serializable
data class ConnectBrowserResponse(
    val success: Boolean,
    val browserType: String,
    val debugPort: Int,
    val wsEndpoint: String? = null,
    val message: String? = null
)

@Serializable
data class DisconnectBrowserResponse(
    val success: Boolean,
    val message: String? = null
)

@Serializable
data class BrowserConnectionStatus(
    val success: Boolean = true,
    val connected: Boolean,
    val browserType: String? = null,
    val debugPort: Int? = null,
    val tabCount: Int? = null,
    val connectTime: Long? = null,
    val commandCount: Int? = null
)

@Serializable
data class UserBrowserTabsResponse(
    val success: Boolean,
    val tabs: List<UserBrowserTab>,
    val total: Int
)

@Serializable
data class UserBrowserTab(
    val targetId: String,
    val url: String,
    val title: String? = null,
    val type: String,
    val attached: Boolean = false,
    val active: Boolean = false,
    val favIconUrl: String? = null
)

@Serializable
data class UserBrowserTabResponse(
    val success: Boolean,
    val tab: UserBrowserTab? = null,
    val message: String? = null
)

@Serializable
data class CloseTabResult(
    val success: Boolean,
    val targetId: String? = null,
    val message: String? = null
)

@Serializable
data class FocusTabResult(
    val success: Boolean,
    val targetId: String? = null,
    val message: String? = null
)

@Serializable
data class NavigateResult(
    val success: Boolean,
    val targetId: String? = null,
    val url: String? = null,
    val frameId: String? = null,
    val loaderId: String? = null,
    val errorText: String? = null
)

@Serializable
data class NavigationResult(
    val success: Boolean,
    val targetId: String? = null,
    val message: String? = null
)

@Serializable
data class ScriptExecutionResult(
    val success: Boolean,
    val result: String? = null,
    val type: String? = null,
    val error: String? = null,
    val exceptionDetails: String? = null
)

@Serializable
data class PageContentResult(
    val success: Boolean,
    val content: String? = null,
    val format: String? = null,
    val url: String? = null,
    val title: String? = null,
    val length: Int? = null
)

@Serializable
data class UserBrowserScreenshotResult(
    val success: Boolean,
    val data: String,  // Base64 编码的图片
    val format: String,
    val width: Int? = null,
    val height: Int? = null,
    val size: Int? = null
)

@Serializable
data class UserBrowserBookmarksResponse(
    val success: Boolean,
    val bookmarks: List<UserBrowserBookmark>,
    val total: Int
)

@Serializable
data class UserBrowserBookmark(
    val id: String,
    val title: String,
    val url: String? = null,
    val parentId: String? = null,
    val dateAdded: Long? = null,
    val dateGroupModified: Long? = null,
    val children: List<UserBrowserBookmark>? = null,
    val type: String = "bookmark"  // bookmark 或 folder
)

@Serializable
data class UserBrowserHistoryResponse(
    val success: Boolean,
    val items: List<UserBrowserHistoryItem>,
    val total: Int
)

@Serializable
data class UserBrowserHistoryItem(
    val id: String,
    val url: String,
    val title: String? = null,
    val lastVisitTime: Long,
    val visitCount: Int,
    val typedCount: Int? = null
)

File: ios-app/Modules/CoreP2P/Sources/RemoteSkills/UserBrowser/UserBrowserCommands.swift
import Foundation

/// 用户浏览器（CDP）控制 typed RPC wrapper — Phase 6.1B1。
///
/// 镜像 Android `app/src/main/java/.../remote/commands/UserBrowserCommands.kt`。
/// 18 method 全 wired，桌面 `user-browser-handler.js` 与 Android 100% 名称对齐
/// （Coverage doc §1.4：A=18 D=18 ✓=18）。
///
/// 通过 Chrome DevTools Protocol (CDP) 连接和控制用户已安装的真实浏览器
/// （Chrome / Edge / Brave / Chromium）。与 `extension` skill 区别：
/// - `extension` = Chrome 扩展独立 WS 子系统（chainlesschain extension 注入）
/// - `userBrowser` = CDP 协议直连用户已装浏览器（无需安装 chainlesschain extension）
///
/// **wire 协议**（与桌面 `user-browser-handler.js` 对齐）：
/// 5 大类 18 method：
/// 1. 连接：`findBrowsers` / `connect` / `disconnect` / `getStatus`
/// 2. 标签页：`listTabs` / `getActiveTab` / `createTab` / `closeTab` / `focusTab`
/// 3. 导航：`navigate` / `goBack` / `goForward` / `refresh`
/// 4. 页面操作：`executeScript` / `getPageContent` / `screenshot`
/// 5. 用户数据：`getBookmarks` / `getHistory`
public actor UserBrowserCommands {

    private let client: RemoteCommandClient

    public init(client: RemoteCommandClient) {
        self.client = client
    }

    // MARK: - 连接管理

    /// 查找可用浏览器（扫桌面已装 Chrome / Edge / Brave 等）。
    public func findBrowsers(
        pcPeerId: String,
        mobileDid: String? = nil
    ) async throws -> FindBrowsersResponse {
        let resp = try await client.invoke(
            pcPeerId: pcPeerId,
            method: "userBrowser.findBrowsers",
            params: [:],
            mobileDid: mobileDid
        )
        return try Self.decode(resp, FindBrowsersResponse.decode)
    }

    /// 连接到指定浏览器（mutating；可能 autoLaunch）。
    public func connect(
        pcPeerId: String,
        browserType: String = "chrome",
        port: Int? = nil,
        autoLaunch: Bool = true,
        mobileDid: String? = nil
    ) async throws -> ConnectBrowserResponse {
        guard !browserType.isEmpty else {
            throw RemoteSkillError.invalidArgument("userBrowser.connect: browserType empty")
        }
        var params: [String: Any] = [
            "browserType": browserType,
            "autoLaunch": autoLaunch
        ]
        if let p = port { params["port"] = p }
        let resp = try await client.invoke(
            pcPeerId: pcPeerId,
            method: "userBrowser.connect",
            params: params,
            mobileDid: mobileDid
        )
        return try Self.decode(resp, ConnectBrowserResponse.decode)
    }

    /// 断开 CDP 连接。
    public func disconnect(
        pcPeerId: String,
        mobileDid: String? = nil
    ) async throws -> DisconnectBrowserResponse {
        let resp = try await client.invoke(
            pcPeerId: pcPeerId,
            method: "userBrowser.disconnect",
            params: [:],
            mobileDid: mobileDid
        )
        return try Self.decode(resp, DisconnectBrowserResponse.decode)
    }

    /// 连接状态。
    public func getStatus(
        pcPeerId: String,
        mobileDid: String? = nil
    ) async throws -> BrowserConnectionStatus {
        let resp = try await client.invoke(
            pcPeerId: pcPeerId,
            method: "userBrowser.getStatus",
            params: [:],
            mobileDid: mobileDid
        )
        return try Self.decode(resp, BrowserConnectionStatus.decode)
    }

    // MARK: - 标签页

    /// 列所有标签页。
    public func listTabs(
        pcPeerId: String,
        mobileDid: String? = nil
    ) async throws -> UserBrowserTabsResponse {
        let resp = try await client.invoke(
            pcPeerId: pcPeerId,
            method: "userBrowser.listTabs",
            params: [:],
            mobileDid: mobileDid
        )
        return try Self.decode(resp, UserBrowserTabsResponse.decode)
    }

    /// 获取当前活动标签页。
    public func getActiveTab(
        pcPeerId: String,
        mobileDid: String? = nil
    ) async throws -> UserBrowserTabResponse {
        let resp = try await client.invoke(
            pcPeerId: pcPeerId,
            method: "userBrowser.getActiveTab",
            params: [:],
            mobileDid: mobileDid
        )
        return try Self.decode(resp, UserBrowserTabResponse.decode)
    }

    /// 创建新标签页（mutating）。
    public func createTab(
        pcPeerId: String,
        url: String? = nil,
        active: Bool = true,
        mobileDid: String? = nil
    ) async throws -> UserBrowserTabResponse {
        var params: [String: Any] = ["active": active]
        if let u = url { params["url"] = u }
        let resp = try await client.invoke(
            pcPeerId: pcPeerId,
            method: "userBrowser.createTab",
            params: params,
            mobileDid: mobileDid
        )
        return try Self.decode(resp, UserBrowserTabResponse.decode)
    }

    /// 关闭标签页（mutating）。
    public func closeTab(
        pcPeerId: String,
        targetId: String,
        mobileDid: String? = nil
    ) async throws -> CloseTabResult {
        guard !targetId.isEmpty else {
            throw RemoteSkillError.invalidArgument("userBrowser.closeTab: targetId empty")
        }
        let resp = try await client.invoke(
            pcPeerId: pcPeerId,
            method: "userBrowser.closeTab",
            params: ["targetId": targetId],
            mobileDid: mobileDid
        )
        return try Self.decode(resp, CloseTabResult.decode)
    }

    /// 聚焦标签页（mutating UI 状态）。
    public func focusTab(
        pcPeerId: String,
        targetId: String,
        mobileDid: String? = nil
    ) async throws -> FocusTabResult {
        guard !targetId.isEmpty else {
            throw RemoteSkillError.invalidArgument("userBrowser.focusTab: targetId empty")
        }
        let resp = try await client.invoke(
            pcPeerId: pcPeerId,
            method: "userBrowser.focusTab",
            params: ["targetId": targetId],
            mobileDid: mobileDid
        )
        return try Self.decode(resp, FocusTabResult.decode)
    }

    // MARK: - 导航

    /// 导航到 URL（mutating）。
    public func navigate(
        pcPeerId: String,
        targetId: String,
        url: String,
        mobileDid: String? = nil
    ) async throws -> NavigateResult {
        guard !targetId.isEmpty else {
            throw RemoteSkillError.invalidArgument("userBrowser.navigate: targetId empty")
        }
        guard !url.isEmpty else {
            throw RemoteSkillError.invalidArgument("userBrowser.navigate: url empty")
        }
        let resp = try await client.invoke(
            pcPeerId: pcPeerId,
            method: "userBrowser.navigate",
            params: ["targetId": targetId, "url": url],
            mobileDid: mobileDid
        )
        return try Self.decode(resp, NavigateResult.decode)
    }

    /// 后退。
    public func goBack(
        pcPeerId: String,
        targetId: String,
        mobileDid: String? = nil
    ) async throws -> NavigationResult {
        guard !targetId.isEmpty else {
            throw RemoteSkillError.invalidArgument("userBrowser.goBack: targetId empty")
        }
        let resp = try await client.invoke(
            pcPeerId: pcPeerId,
            method: "userBrowser.goBack",
            params: ["targetId": targetId],
            mobileDid: mobileDid
        )
        return try Self.decode(resp, NavigationResult.decode)
    }

    /// 前进。
    public func goForward(
        pcPeerId: String,
        targetId: String,
        mobileDid: String? = nil
    ) async throws -> NavigationResult {
        guard !targetId.isEmpty else {
            throw RemoteSkillError.invalidArgument("userBrowser.goForward: targetId empty")
        }
        let resp = try await client.invoke(
            pcPeerId: pcPeerId,
            method: "userBrowser.goForward",
            params: ["targetId": targetId],
            mobileDid: mobileDid
        )
        return try Self.decode(resp, NavigationResult.decode)
    }

    /// 刷新页面。
    public func refresh(
        pcPeerId: String,
        targetId: String,
        ignoreCache: Bool = false,
        mobileDid: String? = nil
    ) async throws -> NavigationResult {
        guard !targetId.isEmpty else {
            throw RemoteSkillError.invalidArgument("userBrowser.refresh: targetId empty")
        }
        let resp = try await client.invoke(
            pcPeerId: pcPeerId,
            method: "userBrowser.refresh",
            params: ["targetId": targetId, "ignoreCache": ignoreCache],
            mobileDid: mobileDid
        )
        return try Self.decode(resp, NavigationResult.decode)
    }

    // MARK: - 页面操作

    /// 执行 JavaScript（高危 Admin；应配 ApprovalGate）。
    public func executeScript(
        pcPeerId: String,
        targetId: String,
        script: String,
        mobileDid: String? = nil
    ) async throws -> ScriptExecutionResult {
        guard !targetId.isEmpty else {
            throw RemoteSkillError.invalidArgument("userBrowser.executeScript: targetId empty")
        }
        guard !script.isEmpty else {
            throw RemoteSkillError.invalidArgument("userBrowser.executeScript: script empty")
        }
        let resp = try await client.invoke(
            pcPeerId: pcPeerId,
            method: "userBrowser.executeScript",
            params: ["targetId": targetId, "script": script],
            mobileDid: mobileDid
        )
        return try Self.decode(resp, ScriptExecutionResult.decode)
    }

    /// 获取页面内容（html / text）。
    public func getPageContent(
        pcPeerId: String,
        targetId: String,
        format: String = "html",
        mobileDid: String? = nil
    ) async throws -> PageContentResult {
        guard !targetId.isEmpty else {
            throw RemoteSkillError.invalidArgument("userBrowser.getPageContent: targetId empty")
        }
        guard format == "html" || format == "text" else {
            throw RemoteSkillError.invalidArgument("userBrowser.getPageContent: format must be html/text")
        }
        let resp = try await client.invoke(
            pcPeerId: pcPeerId,
            method: "userBrowser.getPageContent",
            params: ["targetId": targetId, "format": format],
            mobileDid: mobileDid
        )
        return try Self.decode(resp, PageContentResult.decode)
    }

    /// 截图（与 `display.screenshot` 不同：此为 tab 级；`display` 为整屏）。
    public func screenshot(
        pcPeerId: String,
        targetId: String,
        format: String = "png",
        quality: Int? = nil,
        fullPage: Bool = false,
        mobileDid: String? = nil
    ) async throws -> UserBrowserScreenshotResult {
        guard !targetId.isEmpty else {
            throw RemoteSkillError.invalidArgument("userBrowser.screenshot: targetId empty")
        }
        guard format == "png" || format == "jpeg" || format == "jpg" else {
            throw RemoteSkillError.invalidArgument("userBrowser.screenshot: format must be png/jpeg")
        }
        var params: [String: Any] = [
            "targetId": targetId,
            "format": format,
            "fullPage": fullPage
        ]
        if let q = quality {
            guard q >= 1 && q <= 100 else {
                throw RemoteSkillError.invalidArgument("userBrowser.screenshot: quality must be 1-100")
            }
            params["quality"] = q
        }
        let resp = try await client.invoke(
            pcPeerId: pcPeerId,
            method: "userBrowser.screenshot",
            params: params,
            mobileDid: mobileDid
        )
        return try Self.decode(resp, UserBrowserScreenshotResult.decode)
    }

    // MARK: - 用户数据

    /// 获取书签（可按 folderId 过滤）。
    public func getBookmarks(
        pcPeerId: String,
        folderId: String? = nil,
        mobileDid: String? = nil
    ) async throws -> UserBrowserBookmarksResponse {
        var params: [String: Any] = [:]
        if let f = folderId { params["folderId"] = f }
        let resp = try await client.invoke(
            pcPeerId: pcPeerId,
            method: "userBrowser.getBookmarks",
            params: params,
            mobileDid: mobileDid
        )
        return try Self.decode(resp, UserBrowserBookmarksResponse.decode)
    }

    /// 获取历史记录（可按 query / 时间范围 过滤）。
    public func getHistory(
        pcPeerId: String,
        query: String? = nil,
        maxResults: Int = 100,
        startTime: Int64? = nil,
        endTime: Int64? = nil,
        mobileDid: String? = nil
    ) async throws -> UserBrowserHistoryResponse {
        guard maxResults > 0 else {
            throw RemoteSkillError.invalidArgument("userBrowser.getHistory: maxResults must be > 0")
        }
        var params: [String: Any] = ["maxResults": maxResults]
        if let q = query { params["query"] = q }
        if let s = startTime { params["startTime"] = s }
        if let e = endTime { params["endTime"] = e }
        let resp = try await client.invoke(
            pcPeerId: pcPeerId,
            method: "userBrowser.getHistory",
            params: params,
            mobileDid: mobileDid
        )
        return try Self.decode(resp, UserBrowserHistoryResponse.decode)
    }

    private static func decode<T>(
        _ resp: TerminalRpcResponse,
        _ decoder: (String) throws -> T
    ) throws -> T {
        switch resp {
        case .success(_, let resultJson):
            return try decoder(resultJson)
        case .failure(let reqId, let msg):
            throw RemoteSkillError.remoteError(reqId: reqId, message: msg)
        }
    }
}
